### DISCUSSION TITLE
v4.0.0-beta.1

### DISCUSSION THREAD

**Discussion by PaulRBerg** [PaulRBerg/prb-proxy#77] [Announcements]
### Changed

- Change license to MIT ([#49](https://github.com/PaulRBerg/prb-proxy/issues/49)) (@PaulRBerg)
- Improve documentation (@PaulRBerg)
- Improve formatting (@PaulRBerg)
- Improve names of custom errors, events, and functions (@PaulRBerg)
- Lower pragma to `>=0.8.4` in interface files ([46a34c](https://github.com/PaulRBerg/prb-proxy/commit/46a34c))
  (@PaulRBerg)
- Mark factory as immutable ([bcc8aa](https://github.com/PaulRBerg/prb-proxy/commit/bcc8aa)) (@PaulRBerg)
- Move interface files in a nested directory ([b954b5](https://github.com/PaulRBerg/prb-proxy/commit/b954b5))
  (@PaulRBerg)
- Move the `setPermission` function to an enshrined target contract
  ([3f5794](https://github.com/PaulRBerg/prb-proxy/commit/3f5794)) (@PaulRBerg)
- Reduce deployment size by setting optimizer runs to 200
  ([c2f955](https://github.com/PaulRBerg/prb-proxy/commit/c2f955)) (@PaulRBerg)
- Rename custom errors to use single underscore ([b954b5](https://github.com/PaulRBerg/prb-proxy/commit/b954b5))
  (@PaulRBerg)
- Simplify the envoy permission system ([#72](https://github.com/PaulRBerg/prb-proxy/issues/73)) (@PaulRBerg)
- Update pragmas (@PaulRBerg)
- Use named arguments in function calls (@PaulRBerg)

### Added

- Add `deployAndExecute` functions ([ce9d29](https://github.com/PaulRBerg/prb-proxy/commit/ce9d29)) (@PaulRBerg)
- Add enshrined target contract `PRBProxyHelpers` ([4ca1c9](https://github.com/PaulRBerg/prb-proxy/commit/4ca1c9),
  [3f5794](https://github.com/PaulRBerg/prb-proxy/commit/3f5794)) (@PaulRBerg)
- Add plugin system and fallback function ([#53](https://github.com/PaulRBerg/prb-proxy/pull/53),
  [4ca1c9](https://github.com/PaulRBerg/prb-proxy/commit/4ca1c9)) (@cleanunicorn, @PaulRBerg)
- Add storage contract that replicates the storage layout of the proxy
  ([1449da](https://github.com/PaulRBerg/prb-proxy/commit/1449da)) (@PaulRBerg)
- Emit event in `setPermission` ([b277fd](https://github.com/PaulRBerg/prb-proxy/commit/b277fd)) (@PaulRBerg)
- Re-implement the `setMinGasReserve` function in the enshrined target
  ([e6fdfb](https://github.com/PaulRBerg/prb-proxy/commit/e6fdfb)) (@PaulRBerg)

### Fixed

- Fix the calldata load bug that permitted unintentional calls to fallback functions
  ([#50](https://github.com/PaulRBerg/prb-proxy/issues/50)) (@PaulRBerg)

<hr /><em>This discussion was created from the release <a href='https://github.com/PaulRBerg/prb-proxy/releases/tag/4.0.0-beta.1'>4.0.0-beta.1</a>.</em>

**Comment by PaulRBerg** [comment]
## Feedback

I would appreciate it if I received feedback on this pre-release from @SolcAndMe, @razgraf, and @andreivladbrg. These are the key areas of interest:

1. Deployment cost (my estimate is [528,529](https://goerli.etherscan.io/tx/0xff910d05e22300982794b8eab52434e588530c70afafc6e837d0729e96fbac7f), which would cost ~$27.16 at the current base fee of ~29 gwei on Ethereum). This is 6% more gas efficient compared to V3.
    - The `execute` function has remained as gas efficient as before.
3. User experience of handling the enshrined target `PRBProxyHelpers` (https://github.com/PaulRBerg/prb-proxy/issues/74)
4. Correctness of the newly added `deployAndExecute` functions (https://github.com/PaulRBerg/prb-proxy/issues/62)
5. (Obviously) Any bugs, errors, or things that could be improved

## Contract Addresses

| Contract | Address                                                                                                                           |
| -------- | --------------------------------------------------------------------------------------------------------------------------------- |
| Registry | [0x7aE53f08B444D481d207e5F335Df017CE6E9d248](https://goerli.etherscan.io/address/0x7aE53f08B444D481d207e5F335Df017CE6E9d248#code) |
| Helpers  | [0xDc608dFCc62E0e7756439270D09881EEDf7885EF](https://goerli.etherscan.io/address/0xDc608dFCc62E0e7756439270D09881EEDf7885EF#code) |

**Reply by SolcAndMe** [reply]
Thanks! I don't have much to add. We confirmed our user flow of deploying the proxy, installing plugins, executing an action all in the same transaction via `deployAndExecute`.

**Reply by PaulRBerg** [reply]
Fantastic! Thanks for confirming, @SolcAndMe.

I am still due to make a decision about #79 and #80. Once I do so, and I implement any resultant changes, I will ship v4.

**Comment by PaulRBerg** [comment]
Locking since this is no longer the most up-to-date version. [`v4.0.0-beta.2`](https://github.com/PaulRBerg/prb-proxy/discussions/82) is.
